File: algorithms/587. Erect the Fence/solution.py
# Definition for a point.
# class Point(object):
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution(object):
    def outerTrees(self, points):
        """
        :type points: List[Point]
        :rtype: List[Point]
        """
        # @awice
        def clockwise(p, q, r):
            return (p.x - r.x) * (q.y - r.y) < (q.x - r.x) * (p.y - r.y)
        def findhull(hull, p):
            hull.append(p)
            while len(hull) > 2 and clockwise(*hull[-3:]):
                hull.pop(-2)
            return hull
        points.sort(key=lambda p: (p.x, p.y))
        lowerhull = reduce(findhull, points, [])
        upperhull = reduce(findhull, points[::-1], [])
        return list(set(lowerhull + upperhull))

File: algorithms/587. Erect the Fence/solution.cc
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
bool operator < (const Point & p, const Point & q) {
    return tie(p.x, p.y) < tie(q.x, q.y);
}

class Solution {
public:
    vector<Point> outerTrees(vector<Point>& points) {
        sort(points.begin(), points.end());
        vector<Point> lowerhull;
        for ( auto & p : points ) {
            findhull(p, lowerhull);
        }
        reverse(points.begin(), points.end());
        vector<Point> upperhull;
        for ( auto & p : points ) {
            findhull(p, upperhull);
        }
        lowerhull.insert(lowerhull.end(), upperhull.begin(), upperhull.end());
        set<Point> s(lowerhull.begin(), lowerhull.end());
        return vector<Point> (s.begin(), s.end());
    }

    bool clockwise(Point & p, Point & q, Point & r) {
        return (p.x - r.x) * (q.y - r.y) < (p.y - r.y) * (q.x - r.x);
    }

    void findhull(Point & r, vector<Point> & hull) {
        while ( hull.size() > 1 && clockwise(hull[hull.size()-2], hull.back(), r) ) {
            hull.pop_back();
        }
        hull.push_back(r);
    }
};
